# 3.8. Yürütme Mesajları

Instantiate ve query mesajlarını inceledik. Sonunda son temel giriş noktası olan yürütme mesajlarını tanıtmanın zamanı geldi. Şimdiye kadar yaptıklarımıza benzer şekilde, bunun sadece ürpertici olmasını ve bilgilerimizi tekrar gözden geçirmesini bekliyorum. Burada anlattıklarımı kaynak koduna bakmadan kendi başınıza bir alıştırma olarak uygulamayı denemenizi tavsiye ederim.

Kontrat fikri kolay olacaktır - her kontrat yöneticisi iki yürütme mesajını çağırmaya uygun olacaktır:

* `AddMembers` mesajı, admin'in admin listesine başka bir adres eklemesine izin verecektir
* `Leave`, adminlerin kendilerini listeden çıkarmalarına izin verecektir

Çok karmaşık değil. Hadi kodlamaya geçelim. Mesajları tanımlamakla başlayın:

```
#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
pub enum ExecuteMsg {
    AddMembers { admins: Vec<String> },
    Leave {},
}
```

Ve giriş noktasını uygulayın:

```
use crate::msg::{AdminsListResp, ExecuteMsg, GreetResp, InstantiateMsg, QueryMsg};
 
#[allow(dead_code)]
pub fn execute(
    deps: DepsMut,
    _env: Env,
    info: MessageInfo,
    msg: ExecuteMsg,
) -> StdResult<Response> {
    use ExecuteMsg::*;

    match msg {
        AddMembers { admins } => exec::add_members(deps, info, admins),
        Leave {} => exec::leave(deps, info),
    }
}

mod exec {
    use cosmwasm_std::StdError;

    use super::*;

    pub fn add_members(
        deps: DepsMut,
        info: MessageInfo,
        admins: Vec<String>,
    ) -> StdResult<Response> {
        let mut curr_admins = ADMINS.load(deps.storage)?;
        if !curr_admins.contains(&info.sender) {
            return Err(StdError::generic_err("Unauthorised access"));
        }

        let admins: StdResult<Vec<_>> = admins
            .into_iter()
            .map(|addr| deps.api.addr_validate(&addr))
            .collect();

        curr_admins.append(&mut admins?);
        ADMINS.save(deps.storage, &curr_admins)?;

        Ok(Response::new())
    }

    pub fn leave(deps: DepsMut, info: MessageInfo) -> StdResult<Response> {
        ADMINS.update(deps.storage, move |admins| -> StdResult<_> {
            let admins = admins
                .into_iter()
                .filter(|admin| *admin != info.sender)
                .collect();
            Ok(admins)
        })?;

        Ok(Response::new())
    }
}
```

Giriş noktasının kendisi de `src/lib.rs` içinde oluşturulmalıdır:

```
use cosmwasm_std::{entry_point, Binary, Deps, DepsMut, Env, MessageInfo, Response, StdResult};
use msg::{ExecuteMsg, InstantiateMsg, QueryMsg};

mod contract;
mod msg;
mod state;

#[entry_point]
pub fn instantiate(
    deps: DepsMut,
    env: Env,
    info: MessageInfo,
    msg: InstantiateMsg,
) -> StdResult<Response> {
    contract::instantiate(deps, env, info, msg)
}

#[entry_point]
pub fn execute(deps: DepsMut, env: Env, info: MessageInfo, msg: ExecuteMsg) -> StdResult<Response> {
    contract::execute(deps, env, info, msg)
}

#[entry_point]
pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
    contract::query(deps, env, msg)
}
```

Birkaç yeni şey var ama önemli değil. Birincisi, admin olduğunu doğrulamak veya listeden çıkarmak için mesajı gönderene - `MessageInfo`'nun `info.sender` alanını göründüğü gibi kullandım - nasıl ulaşacağım. Mesaj her zaman doğru adresten gönderildiği için, `sender` zaten `Addr` türündedir - doğrulamaya gerek yoktur. Bir başka yeni şey de bir `Item` üzerindeki [`update`](https://docs.rs/cw-storage-plus/0.13.4/cw\_storage\_plus/struct.Item.html#method.update) işlevidir - bir varlığın okunmasını ve güncellenmesini potansiyel olarak daha verimli hale getirir. Bunu önce adminleri okuyarak, ardından sonucu güncelleyerek ve depolayarak yapmak mümkündür.

Muhtemelen `Item` ile çalışırken her zaman bir şeyin orada olduğunu varsaydığımızı fark etmişsinizdir. Ancak hiçbir şey bizi `ADMINS` değerini başlangıçta başlatmaya zorlamaz! Peki orada ne oluyor? Hem `load` hem de `update` fonksiyonları hata döndürecektir. Ancak `StdResult<Option>` döndüren bir [`may_load`](https://docs.rs/cw-storage-plus/0.13.4/cw\_storage\_plus/struct.Item.html#method.may\_load) fonksiyonu vardır - boş depolama durumunda `Ok(None)` döndürür. Hatta remove fonksiyonu ile mevcut bir öğeyi depodan [`remove`](https://docs.rs/cw-storage-plus/0.13.4/cw\_storage\_plus/struct.Item.html#method.remove) etme imkanı da vardır.

Geliştirilmesi gereken bir şey de hata işleme. Göndericinin yönetici olduğunu doğrularken, hata olarak rastgele bir dize döndürüyoruz. Daha iyisini yapabiliriz.

### Hata işleme

Kontratımızda, bir kullanıcı kendisi bir yönetici olmadan `AddMembers`'ı çalıştırmaya çalıştığında bir hata durumumuz var. [`StdError`](https://docs.rs/cosmwasm-std/1.0.0/cosmwasm\_std/enum.StdError.html)'da bu durumu bildirmek için uygun bir hata durumu yoktur, bu nedenle bir mesajla birlikte genel bir hata döndürmemiz gerekir. Bu en iyi yaklaşım değildir.

Hata raporlama için [`thiserror`](https://crates.io/crates/thiserror/1.0.24/dependencies) crate'i kullanmanızı öneririz. Bağımlılıklarınızı güncelleyerek başlayın:

```
[package]
name = "contract"
version = "0.1.0"
edition = "2021"

[lib]
crate-type = ["cdylib"]

[dependencies]
cosmwasm-std = { version = "1.0.0-beta8", features = ["staking"] }
serde = { version = "1.0.103", default-features = false, features = ["derive"] }
cw-storage-plus = "0.13.4"
thiserror = "1"

[dev-dependencies]
cw-multi-test = "0.13.4"
```

Şimdi `src/error.rs` içinde bir hata türü tanımlıyoruz:

```
use cosmwasm_std::{Addr, StdError};
use thiserror::Error;

#[derive(Error, Debug, PartialEq)]
pub enum ContractError {
    #[error("{0}")]
    StdError(#[from] StdError),
    #[error("{sender} is not contract admin")]
    Unauthorized { sender: Addr },
}
```

Yeni modülü `src/lib.rs` dosyasına da eklememiz gerekiyor:

```
mod contract;
mod error;
mod msg;
mod state;
```

`thiserror` kullanarak hataları basit bir enum gibi tanımlarız ve sandık, türün [`std::error::Error`](https://doc.rust-lang.org/std/error/trait.Error.html) özelliğini uygulamasını sağlar. Bu crate'in çok güzel bir özelliği, [`Display`](https://doc.rust-lang.org/std/fmt/trait.Display.html) özelliğinin bir `#[error]` niteliği ile satır içi tanımlanmasıdır. Ayrıca, bir başka yararlı şey de otomatik olarak uygun [`From`](https://doc.rust-lang.org/std/convert/trait.From.html) uygulamasını oluşturan `#[from]` niteliğidir, böylece `?` operatörünü `thiserror` türleriyle kullanmak kolaydır.

Şimdi yeni hata türümüzü kullanmak için execute uç noktasını güncelleyin:

```
use crate::error::ContractError;
use crate::msg::{AdminsListResp, ExecuteMsg, GreetResp, InstantiateMsg, QueryMsg};
 
pub fn execute(
    deps: DepsMut,
    _env: Env,
    info: MessageInfo,
    msg: ExecuteMsg,
) -> Result<Response, ContractError> {
    use ExecuteMsg::*;

    match msg {
        AddMembers { admins } => exec::add_members(deps, info, admins),
        Leave {} => exec::leave(deps, info).map_err(Into::into),
    }
}

mod exec {
    use super::*;

    pub fn add_members(
        deps: DepsMut,
        info: MessageInfo,
        admins: Vec<String>,
    ) -> Result<Response, ContractError> {
        let mut curr_admins = ADMINS.load(deps.storage)?;
        if !curr_admins.contains(&info.sender) {
            return Err(ContractError::Unauthorized {
                sender: info.sender,
            });
        }

        let admins: StdResult<Vec<_>> = admins
            .into_iter()
            .map(|addr| deps.api.addr_validate(&addr))
            .collect();

        curr_admins.append(&mut admins?);
        ADMINS.save(deps.storage, &curr_admins)?;

        Ok(Response::new())
    }
}
```

Giriş noktası dönüş türünün de güncellenmesi gerekir:

```
use error::ContractError;

#[entry_point]
pub fn execute(
    deps: DepsMut,
    env: Env,
    info: MessageInfo,
    msg: ExecuteMsg,
) -> Result<Response, ContractError> {
    contract::execute(deps, env, info, msg)
}
```

### [Özel hata ve çoklu test](3.8.-yueruetme-mesajlari.md#oezel-hata-ve-coklu-test)

Uygun özel hata türünü kullanmanın güzel bir avantajı vardır - multi-test [`anyhow`](https://crates.io/crates/anyhow) crate kullanarak hata türünü korur. Bu, `thiserror`'ın bir kardeşidir ve orijinal hatayı geri almaya izin verecek şekilde türden silinmiş hataları uygulamak için tasarlanmıştır.

Admin olmayan birinin kendisini bir listeye ekleyemeyeceğini doğrulayan bir test yazalım:

```
#[cfg(test)]
mod tests {
    #[test]
    fn unauthorized() {
        let mut app = App::default();

        let code = ContractWrapper::new(execute, instantiate, query);
        let code_id = app.store_code(Box::new(code));

        let addr = app
            .instantiate_contract(
                code_id,
                Addr::unchecked("owner"),
                &InstantiateMsg { admins: vec![] },
                &[],
                "Contract",
                None,
            )
            .unwrap();

        let err = app
            .execute_contract(
                Addr::unchecked("user"),
                addr,
                &ExecuteMsg::AddMembers {
                    admins: vec!["user".to_owned()],
                },
                &[],
            )
            .unwrap_err();

        assert_eq!(
            ContractError::Unauthorized {
                sender: Addr::unchecked("user")
            },
            err.downcast().unwrap()
        );
    }
}
```

Bir kontratın yürütülmesi diğer çağrılara çok benzer - bir [`execute_contract`](https://docs.rs/cw-multi-test/0.13.4/cw\_multi\_test/trait.Executor.html#method.execute\_contract) fonksiyonu kullanırız. Yürütme başarısız olabileceğinden, bu çağrıdan bir hata türü alırız, ancak bir değer çıkarmak için `unwrap`'ı çağırmak yerine, bir hata oluşmasını bekleriz - [`unwrap_err`](https://doc.rust-lang.org/std/result/enum.Result.html#method.unwrap\_err) çağrısının amacı budur. Şimdi, bir hata değerine sahip olduğumuz için, bir `assert_eq` ile beklediğimizle eşleşip eşleşmediğini kontrol edebiliriz! Küçük bir sorun var - `execute_contract`'tan dönen hata bir [`anyhow::Error`](https://docs.rs/anyhow/1.0.57/anyhow/struct.Error.html) hatasıdır, ancak biz bunun bir `ContractError` olmasını bekliyoruz. Umarım, daha önce de söylediğim gibi, `anyhow` hataları [`downcast`](https://docs.rs/anyhow/1.0.57/anyhow/struct.Error.html#method.downcast) fonksiyonunu kullanarak orijinal türlerini geri kazanabilirler. Hemen ardından gelen `unwrap` gereklidir çünkü downcasting başarısız olabilir. Bunun nedeni, `downcast`'in altta yatan hatada tutulan türü sihirli bir şekilde bilmemesidir. Bunu bir bağlamdan çıkarır - burada, onunla karşılaştırıldığı için bir `ContractError` olmasını beklediğimizi bilir - tip elision mucizeleri. Ancak altta yatan hata bir `ContractError` olmasaydı, `unwrap` paniğe kapılırdı.

Az önce yürütme için basit bir hata testi oluşturduk, ancak kontratın üretime hazır olduğunu iddia etmek yeterli değildir. Bunun için tüm makul tamam durumları kapsanmalıdır. Bu bölümden sonra alıştırma olarak bazı testler oluşturmanızı ve bunları denemenizi tavsiye ederim.

